tools/migration/internal/runner.go

```diff
@@ -27,22 +40,112 @@ type Migration interface {
 }
 
 type MigrationRunner struct {
-	verbose    bool
-	command    string
+	// verbose controls the amount of output.
+	verbose bool
+
+	// command is the migration command to run, passed from the CLI
+	command string
+
+	// oldRepoOpt is the option passed  from the CLI.
 	oldRepoOpt string
+
+	// MigrationsProvider is a dependency for fetching available migrations
+	// to allow unit tests to supply test migrations without creating test fixtures.
+	MigrationsProvider func() []Migration
+
+	// RepoVersionGetter is a dependency for getting a Repo Version,
+	// to allow the binary version of the repo to be mocked in testing.
+	RepoVersionGetter func() uint
 }
 
 func NewMigrationRunner(verb bool, command, oldRepoOpt string) *MigrationRunner {
-	// TODO: Issue #2585 Implement repo migration version detection and upgrade decisioning
-
 	return &MigrationRunner{
-		verbose:    verb,
-		command:    command,
-		oldRepoOpt: oldRepoOpt,
+		verbose:            verb,
+		command:            command,
+		oldRepoOpt:         oldRepoOpt,
+		MigrationsProvider: DefaultMigrationsProvider,
+		RepoVersionGetter:  DefaultVersionGetter,
 	}
 }
 
+// DefaultVersionGetter fetches the Version constant from the repo package.
+func DefaultVersionGetter() uint {
+	return repo.Version
+}
+
 func (m *MigrationRunner) Run() error {
 	// TODO: Issue #2595 Implement first repo migration
+	repoVersion, err := m.loadVersion()
+	if err != nil {
+		return err
+	}
+	if repoVersion == repo.Version {
+		return errors.New("binary version = repo version; migration not run")
+	}
```

Comment:
As above, the repo version at which the migration binary was built should not be relevant. Inspect the repo that was specified as the target (oldRepoOpt)

Reply:
We need some way to figure out which version is the most recent that this binary will try to get us to.  Using repo.Version seems like a good way to do this.  If we want to separate out to prevent dependency shifts from messing up rebuilds then we could have a MaxVersion or something defined in the migration binary.

Reply:
So now we look at our migrations list and find the largest version number.  We don't need to depend on the repo version number in fsrepo.

Reply:
I don't think I understand this comment entirely. The user will definitely want to know if they are up to date as opposed to some other version condition. It should not just say a valid migration can't be found.  The other reason to do this check is an early bail out. If there's nothing to do, then why go through the rigamarole of finding valid migrations. 

Since we know what the end migration version will be from what's in our own list, as you pointed out @anorth, technically it's nothing to do with what is at `repo.Version`, although these had better be the same. I used repo.Version initially, because it was quick to check and because it should never be out of sync.

Failing to find valid migrations otherwise, I think should actually produce an error; please see my comment below.